rust/src/client.rs

```diff
@@ -0,0 +1,842 @@
+    ffi::{self, CStr, CString},
+    os::raw::{c_char, c_void},
+    ptr, slice,
+    time::Duration,
+};
+
+use crossbeam_utils::thread;
+use libsettings_sys::{
+    sbp_msg_callback_t, sbp_msg_callbacks_node_t, settings_api_t, settings_create,
+    settings_destroy, settings_read_bool, settings_read_by_idx, settings_read_float,
+    settings_read_int, settings_read_str, settings_t,
+    settings_write_res_e_SETTINGS_WR_MODIFY_DISABLED, settings_write_res_e_SETTINGS_WR_OK,
+    settings_write_res_e_SETTINGS_WR_PARSE_FAILED, settings_write_res_e_SETTINGS_WR_READ_ONLY,
+    settings_write_res_e_SETTINGS_WR_SERVICE_FAILED,
+    settings_write_res_e_SETTINGS_WR_SETTING_REJECTED, settings_write_res_e_SETTINGS_WR_TIMEOUT,
+    settings_write_res_e_SETTINGS_WR_VALUE_REJECTED, settings_write_str, size_t,
+};
+use log::{debug, error};
+use sbp::{
+    link::{Key, Link},
+    messages::{SBPMessage, SBP},
+    serialize::SbpSerialize,
+};
+
+use crate::{settings, SettingKind, SettingValue};
+
+const SENDER_ID: u16 = 66;
+
+pub struct Client<'a> {
+    inner: Box<ClientInner<'a>>,
+}
+
+impl<'a> Client<'a> {
+    pub fn new<F>(link: Link<'a, ()>, sender: F) -> Client<'a>
+    where
+        F: FnMut(SBP) -> Result<(), Box<dyn std::error::Error + Send + Sync>> + 'static,
+    {
+        let context = Box::new(Context {
+            link,
+            sender: Box::new(sender),
+            callbacks: Vec::new(),
+            lock: Lock::new(),
+            event: None,
+        });
+
+        let api = Box::new(settings_api_t {
+            ctx: ptr::null_mut(),
+            send: Some(libsettings_send),
+            send_from: Some(libsettings_send_from),
+            wait_init: Some(libsettings_wait_init),
+            wait: Some(libsettings_wait),
+            wait_deinit: None,
+            signal: Some(libsettings_signal),
+            wait_thd: Some(libsettings_wait_thd),
+            signal_thd: Some(libsettings_signal_thd),
+            lock: Some(libsettings_lock),
+            unlock: Some(libsettings_unlock),
+            register_cb: Some(register_cb),
+            unregister_cb: Some(unregister_cb),
+            log: None,
+            log_preformat: false,
+        });
+
+        let mut inner = Box::new(ClientInner {
+            context: ptr::null_mut(),
+            api: Box::into_raw(api),
+            ctx: ptr::null_mut(),
+        });
+
+        let context_raw = Box::into_raw(context);
+        inner.context = context_raw;
+
+        // Safety: inner.api was just created via Box::into_raw so it is
+        // properly aligned and non-null
+        unsafe {
+            (*inner.api).ctx = context_raw as *mut c_void;
+        }
+
+        inner.ctx = unsafe { settings_create(SENDER_ID, inner.api) };
+        assert!(!inner.ctx.is_null());
+
+        Client { inner }
+    }
+
+    pub fn read_all(&self) -> Vec<Result<ReadByIdxResult, Error<ReadSettingError>>> {
+        const NUM_WORKERS: usize = 5;
+
+        thread::scope(move |scope| {
```

Comment:
could have used rayon here but from my understanding it's mostly intended for cpu bound task where this is more io? Plus I couldn't get the parallel iterators to work, seems like they need to have a known size maybe? Tried something like `(0..).into_par_iter().map(..).while_some()` but open ranges don't implement par iter. Either way managed to reuse the same worker threads in each call to read_all

Reply:
I think avoiding rayon is fine since (as you say) it is mostly intended for compute intensive things and it's another big dependency.  This is a nice solution using a channel to send the threads work.